retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/RetrofitClassAwareMeterIdPrefixFunction.java

```diff
@@ -0,0 +1,172 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.logging.RequestLog;
+import com.linecorp.armeria.common.metric.MeterIdPrefix;
+import com.linecorp.armeria.internal.metric.RequestMetricSupport;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import retrofit2.Invocation;
+import retrofit2.http.DELETE;
+import retrofit2.http.GET;
+import retrofit2.http.HEAD;
+import retrofit2.http.HTTP;
+import retrofit2.http.OPTIONS;
+import retrofit2.http.PATCH;
+import retrofit2.http.POST;
+import retrofit2.http.PUT;
+
+/**
+ * Returns the default function for retrofit that creates a {@link MeterIdPrefix} with the specified name and
+ * the {@link Tag}s derived from the {@link RequestLog} properties, {@link Invocation} and provided retrofit
+ * client.
+ * <ul>
+ *     <li>{@code service} (or {code serviceTagName}) - Retrofit service interface name</li>
+ *     <li>{@code path}   - Retrofit service interface method path taken from method annotation
+ *                          or {@code UNKNOWN} if Retrofit service interface method available</li>
+ *     <li>{@code method} - Retrofit service interface method
+ *                          or {@code UNKNOWN} if Retrofit service interface method available</li>
+ *     <li>{@code httpMethod} - HTTP method name from Retrofit service interface method annotation
+ *                              or from {@link RequestLog#method()} if Retrofit service interface
+ *                              method not available</li>
+ *     <li>{@code httpStatus} - {@link HttpStatus#code()}</li>
+ * </ul>
+ */
+final class RetrofitClassAwareMeterIdPrefixFunction extends RetrofitMeterIdPrefixFunction {
+
+    private static final List<Class<?>> RETROFIT_ANNOTATIONS = ImmutableList.of(
+            POST.class, PUT.class, PATCH.class, HEAD.class, GET.class, OPTIONS.class, HTTP.class, DELETE.class
+    );
+    private static final String METHOD_TAG_NAME = "method";
+    private static final String PATH_TAG_NAME = "path";
+    private static final String HTTP_METHOD_TAG_NAME = "httpMethod";
+    private static final List<Tag> DEFAULT_METHOD_TAGS = ImmutableList.of(
+            Tag.of(METHOD_TAG_NAME, "UNKNOWN"),
+            Tag.of(PATH_TAG_NAME, "UNKNOWN")
+    );
+
+    private final Map<String, List<Tag>> methodNameToTags;
+    private final String name;
+    private final String serviceName;
+    private final String serviceTagName;
+
+    RetrofitClassAwareMeterIdPrefixFunction(String name,
+                                            @Nullable String serviceTagName,
+                                            Class<?> serviceClass) {
+        super(name, null, null);
+
+        this.name = name;
+        this.serviceTagName = firstNonNull(serviceTagName, "service");
+        serviceName = serviceClass.getSimpleName();
```

Comment:
Is it possible for a user to use an anonymous class as a service class? In that case, `getSimpleName` will return an empty string. Perhaps we should strip only the package names in that case?

Reply:
Is it possible to create Retrofit client with anonymous class?
May I ask to provide example?

Reply:
It seems like `Retrofit.create(Class)` requires the class to be an interface. I guess it's not possible to create a Retrofit client with an anonymous class, because an anonymous class is never an interface. :sweat_smile: